### PR TITLE
Fix scheduler reporting

### PR DESCRIPTION
**Pull request by muellerzr**
# What does this PR do?

Fixes the issue of the wrong iterations scheduler being reported and adds a test

Solves https://github.com/huggingface/transformers/issues/28124

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@amyeroberts

### CODE REVIEW

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -2255,17 +2255,21 @@ def _inner_training_loop(
                     # Optimizer step
                     self.optimizer.step()
                     optimizer_was_run = not self.accelerator.optimizer_step_was_skipped
+                    # Store the current LR before stepping
+                    learning_rate = self._get_learning_rate()
```
Comment (muellerzr) [inline src/transformers/trainer.py:2259]: I had to modify the order slightly, as we need to grab the LR *before* scheduling, but we need to `step()` it before calling `on_step_end`

File: src/transformers/trainer.py [hunk 2]
```diff
@@ -2255,17 +2255,21 @@ def _inner_training_loop(
...
-
                     model.zero_grad()
                     self.state.global_step += 1
                     self.state.epoch = epoch + (step + 1 + steps_skipped) / steps_in_epoch
                     self.control = self.callback_handler.on_step_end(args, self.state, self.control)
 
-                    self._maybe_log_save_evaluate(tr_loss, grad_norm, model, trial, epoch, ignore_keys_for_eval)
+                    self._maybe_log_save_evaluate(
+                        tr_loss, grad_norm, learning_rate, model, trial, epoch, ignore_keys_for_eval
+                    )
```
Comment (amyeroberts) [inline src/transformers/trainer.py:2271]: Why not move this line above L2260? It seems it's logging a few values before there's any change to the global state e.g. `epoch`

**Review by amyeroberts — COMMENTED** [review]
Thanks for working on this! 

Just some questions so I can better understand the intended logic

File: src/transformers/trainer.py [hunk 3]
```diff
@@ -2285,7 +2289,9 @@ def _inner_training_loop(
                 self.control.should_training_stop = True
 
             self.control = self.callback_handler.on_epoch_end(args, self.state, self.control)
-            self._maybe_log_save_evaluate(tr_loss, grad_norm, model, trial, epoch, ignore_keys_for_eval)
+            self._maybe_log_save_evaluate(
+                tr_loss, grad_norm, self._get_learning_rate(), model, trial, epoch, ignore_keys_for_eval
+            )
```
[the remark this answers is not in the document]
Comment (muellerzr) [reply src/transformers/trainer.py:2294]: First is at `n` steps, the second is at the end of the epoch

File: src/transformers/trainer.py (hunk 2 above)
Comment (muellerzr) [reply src/transformers/trainer.py:2271]: That changes any behavior if users want to modify something with `on_step_end` and fully adjusts when that callback is called. Seems far too reaching for this IMO :) We want that to be a thing

Comment (muellerzr) [reply src/transformers/trainer.py:2271]: (And if we do this, it will make some tests fail IIRC because reported vs expected changes wildly)

Comment (amyeroberts) [reply src/transformers/trainer.py:2271]: If I've understood correctly, this means that the elements being logged aren't all in sync i.e. some will be reporting values before a step and some after

Comment (muellerzr) [reply src/transformers/trainer.py:2271]: Yes, which I think for the scheduler is important because stepping it only reports what the next LR will be when grabbing, which is not what we want. We want the applied scheduler. For the rest not as much I **think**, I'll look back through some eval reports to make sure that it is just an isolated incident.

Comment (amyeroberts) [reply src/transformers/trainer.py:2271]: Sorry, I don't think I've fully understood what should be happening. 

> Yes, which I think for the scheduler is important because stepping it only reports what the next LR will be when grabbing, which is not what we want. We want the applied scheduler. 

Three questions: 
* In terms of what we "want" - do you mean what we want to pass to `maybe_log_save_evaluate` here, or something else?
* What is the difference between the scheduler and the applied scheduler? 
* Why can't we use the applied scheduler

I think it would be good to have @pacman100 review this first, as he's more familiar with all of these objects

> For the rest not as much I think, I'll look back through some eval reports to make sure that it is just an isolated incident.

Were you able to review reports and confirm behaviour is as expected?

**Comment by SunMarc** [comment]
Let's close this PR since the fix was merged in the above PR
